Sprint 6 solution

ds-skoropad opened the pull request:
Sprint 6 solution

ds-skoropad committed:
Sprint 6. Review 1

ds-skoropad committed:
Sprint 6. Review 1

ds-skoropad committed:
Sprint 6. Review 1

ds-skoropad committed:
Sprint 6. Review 1

ds-skoropad committed:
Merge remote-tracking branch 'origin/develop' into develop

ds-skoropad committed:
Sprint 6. Review 1. Final fix.

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/InMemoryHistoryManager.java:
```diff
@@ -2,28 +2,81 @@
+            tasks.add(countNode.getTask());
+            countNode = countNode.getNext();
+        }
+        return tasks;
```
Эти методы вызываются в публичных методах интерфейса, соответственно нет необходимости оставлять их тоже публичными, в них скрыта реализация логики, поэтому их лучше закрыть от внешнего доступа

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/InMemoryHistoryManager.java:
```diff
@@ -2,28 +2,81 @@
+
+    public void linkLast(Task task) { // Добавляет задачу в конец списка
+        if (task == null) return;
+        removeNode(history.get(task.getId()));
```
Метод содержит логику добавления ноды в конец и в нем не логично делать удаление ноды, вызывать этот метод для дедупликации будет логичнее в методе `add()`

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/Node.java:
Класс ноды можно сделать вложенным в `InMemoryHistoryManager` (private static), он не нужен как отдельный класс, потому что используется для реализации истории внутри менеджера истории. Соответственно и не нужны геттеры и сеттеры для доступа к полям, так же как и не нужны доп методы `toString()` и методы сравнения

ds-skoropad replied:
Так намного лучше, спасибо!

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/InMemoryTaskManager.java:
Нужно добавить удаление из истории во все методы удаления в таск менеджере

ds-skoropad replied:
Кошмар, как я это пропустил :(

Ivanov-s-tmn commented on test/ru/yandex/practicum/kanban/manager/InMemoryHistoryManagerTest.java:
Нужно немного переработать тесты, сейчас больше нет ограничений по вместимости, соответственно тесты проверяющие вместимость больше не нужны. А дополнительно нужно реализовать тесты, проверяющие удаление из разных частей истории (из начала, из середины, с конца), важно проверить порядок задач в истории после удаления

ds-skoropad replied:
Это я с ветками перемудрил, при восстановление что-то полетело. Полностью переписал.

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/InMemoryHistoryManager.java:
```diff
@@ -2,28 +2,81 @@
+    public void removeNode(Node node) { // Удаление по Node
+        if (node == null) return;
+        if (history.isEmpty()) return;
+        if (history.size() == 1) {
```
В условиях лучше работать не с длинной истории, а с ссылками, т.е. выстраивать условия исходя из значений ссылок, как пример `if (first == null)` //история пустая
А вообще, получилась достаточно сложная реализация и можно упростить.
Приведу пример реализации, хотя их достаточно много и есть еще более короткие и изящные варианты)
Есть два случая: 
1. У ноды есть ссылка на предыдущую ноду (node.prev != null) т.е. нода может быть как в середине, так и в конце
разрываем связь с удаляемой нодой и присваиваем предыдущей ноде ссылку на следующую за удаляемой нодой:
`node.prev.next = node.next;`
если следующей ноды нет (node.next == null), т.е. наша удаляемая нода в конце, то последней нодой становится прошлая:
`first = node.prev;`
иначе, наша нода находится в середине и у нее есть ссылки на следующую и предыдущую ноду, тогда разрываем связь с удаляемой нодой, только уже ссылками на предыдущую ноду:
`node.next.prev = node.prev;` //теперь следующая нода будет ссылаться не на удаляемую ноду, а на предыдущую за удаляемой 

2. Остальные, т.е. остается ситуация, когда удаляемая нода первая (node == first)
первой нодой становится идущая после удаляемой:
`first = node.next;`
если следующей нет и таким образом first == null, то эта нода была единственной и мы обнуляем последнюю ноду last = null;
в остальных случаях удаляемая нода не единственная и тогда нам надо надо обнулить ссылку на предыдущую ноду новой первой ноды
`first.prev = null;`

ds-skoropad replied:
Мое почтение. Переписал.

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/InMemoryHistoryManager.java:
```diff
@@ -2,28 +2,81 @@
+        } else {
+            last.setNext(newNode);
+            newNode.setPrevious(last);
+            last = newNode;
```
Строчки дублируются в обоих условиях, ее можно вынести. Ноду можно создавать сразу с ссылкой на предыдущую последнюю ноду last

ds-skoropad committed:
Sprint 6. Review 2.

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/InMemoryHistoryManager.java:
```diff
@@ -29,54 +31,54 @@ public List<Task> getHistoryList() { // Возвращает список ист
-            newNode.setPrevious(last);
-            last = newNode;
+            last.next = newNode;
+            newNode.previous = last;
```
Эту ссылку можно дать при создании новой ноды
`Node newNode = new Node(last, task, null);`

ds-skoropad replied:
Мне еще учится и учится )

Ivanov-s-tmn commented on test/ru/yandex/practicum/kanban/manager/InMemoryHistoryManagerTest.java:
```diff
@@ -11,74 +11,132 @@
+        assertEquals(NUM_OF_TASKS - 1, history.size(), "Несоответствие размера списка истории");
+        // Проверяем порядок всех элементов
+        int countId = 2;
+        for (int i = 0; i < history.size(); i++) {
+            assertEquals(countId, history.get(i).getId(), "Несоответствие ID под индексом " + i);
+            countId++;
+        }
```
В тестах стоит избегать циклов. Вообще, будет достаточно 3 тестовых задач для всех тестов. Можно создать конкретные объекты в `BeforeEach` методе без цикла, добавить их в историю, а в самих тестах достаточно 2 строк кода, как пример:
```java
historyManager.remove(1);
assertEquals(List.of(task2, task3), historyManager.getHistoryList());
```

ds-skoropad replied:
Переписал и получил удовольствие :) Спасибо.

ds-skoropad committed:
Sprint 6 review 2

Fix linkLast() in InMemoryHistoryManager
Fix all tests in InMemoryHistoryManagerTest
